implement structured clone callbacks - support Blob cloning

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->
1. Implement stubs for structured clone callbacks.
2. Support Blob cloning.

Partial implementation of https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #15021 (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/15519)
<!-- Reviewable:end -->

gterzian committed:
implement stubs for structured clone callbacks

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/dom/bindings/structuredclone.rs
 * @kichjang: components/script/dom/bindings/structuredclone.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify script code, but no tests are modified. Please consider adding a test!

jdm commented:
It's hard to tell what the purpose of the StructuredCloneCallbackHandler enum is. What do we gain from static methods instead of free functions? I also recommend a static JSStructuredCloneCallbacks value, rather than returning a new instance from get_structured_clone_callbacks.

gterzian commented:
@jdm thank you! I have re-shuffled things around a bit, please take another look.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -7,13 +7,77 @@
+                           r: *mut JSStructuredCloneReader,
+                           tag: u32,
+                           data: u32,
+                           closure: *mut raw::c_void) -> *mut JSObject {
```
nit: indent each subsequent line to match the first argument (in all new functions in this file).

gterzian replied:
ok

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -7,13 +7,77 @@
 use std::slice;
 
+
+#[allow(unused_variables, dead_code)]
```
nit: use `_` prefixes on argument names instead of allowing unused_variables.

gterzian replied:
thanks, done...

jdm reviewed (changes requested):
This looks like what I was expecting to see!

gterzian commented:
@jdm nits have been addressed. Needs a rebase if all OK. Should we do the next step, supporting one concrete use case, in a separate PR?

jdm commented:
I don't see much purpose to merging the changes as-is given that no code makes use of the callbacks.

gterzian commented:
@jdm sure, ok I'll continue in here and you know if I have any questions 😄

gterzian committed:
support structured cloning for Blob

gterzian commented:
I have two questions:

1. In the `read_callback`, we need to identify what type of data we are receiving, based on the `tag`.  In the C++ code, they do that with [StructuredCloneTags](https://dxr.mozilla.org/mozilla-central/source/dom/base/StructuredCloneTags.h#17). However I'm not quite sure how to implement that on our side. It seems they are using [JS_SCTAG_USER_MIN ((uint32_t) 0xFFFF8000)](https://dxr.mozilla.org/mozilla-central/source/js/public/StructuredClone.h#323) as initial value, and then rely on the fact that those enums in C++ will 'accrue' their values based on the first one. 
I did come across this [SCTAG_TMO_USER_MIN](https://github.com/servo/rust-mozjs/blob/master/src/jsapi_macos_64.rs#L2137), but I'm not sure if this value can be used in the same way.

2. In the `write_callback`, we need to determine the type of the `HandleObject`, for example to make sure it is a `Blob`. In C++ they use this [UNWRAP_OBJECT](https://dxr.mozilla.org/mozilla-central/source/dom/bindings/BindingUtils.h#189) which they cast to the appropriate value in this [PrototypeList](https://dxr.mozilla.org/mozilla-central/source/obj-x86_64-pc-linux-gnu/dist/include/mozilla/dom/PrototypeList.h#52). I'm wondering how best to do something similar on our side. Can I use [`Blob`](https://github.com/servo/servo/blob/master/components/script/dom/blob.rs#L71) perhaps? It seems the various objects in `script/dom/` match the prototypes on that list...

Thanks for taking a look.

gterzian commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,101 @@
+                                   _data: u32,
+                                   _closure: *mut raw::c_void) -> *mut JSObject {
+    match tag {
+        tag if tag == StructuredCloneTags::DomBlob as u32 => {
```
this also seems like a terrible hack, any suggestions on a better way to approach this?

jdm commented:
1. When no constant exists, you should add one to https://github.com/servo/rust-mozjs/blob/master/src/consts.rs. You can use the [same enum trick](https://gist.github.com/12cbeb1a95b1d498e70f5306ba4ebc21) in rust, too.
2. Instead of doing the prototype list matching, you can use `root_from_handleobject` which allows passing types like `Blob` as arguments.

gterzian commented:
Ok so after a long fight with the compiler, I've finally gotten something that passes. I'd like to share an update on my "progress", as well as present the choices I've made so far:

* First of all, in the C++ code, they store the clone blobs in an array, I believe held as a static member of the `StructuredCloneHolder`. They then pass this class around the `JS_*` functions as the `closure` argument, and receive it back in the various callbacks. In the `write` callback, they append the blob to the array, and write the index position to the `JSStructuredCloneWriter`. In the `read` callback, the index is received as the `data` argument, and used it to access the blob stored in the array(In fact they store the `BlobImpl`, not the Blob). 
* Interestingly, they do not always use this approach. For other type of cloned data, like images, they don't store anything in the holder, instead completely relying on the various `JS_Read*` and `JS_Write*` functions to pass data around.
* I first tried a similar approach to the blobs in C++. First with a static module level vector, later by passing around a new "holder" struct using the `closure` argument. None of it passed compilation as far as I tried. 
* I then realized we already have the raw data of the object we're cloning, right there in the `StructureCloneData` enum. So I now us a combination of passing the enum around in the `closure` argument, and writing the raw data using `JS_WriteUint32Pair` in the write callback, while just reading it back in the `data` argument of the read callback. (As I understand it, the `u32` pair you write using `JS_WriteUint32Pair`, will be passed to the read callback as `tag` and `data`). 

It now all compiles, although I have to admit a lot of the stuff I did was with only limited understanding. So a review and feedback at this point would be highly appreciated.

gterzian commented:
I also would like to get feedback on this idea for unit testing this stuff:

I've been looking at the [unit test package for script](https://github.com/servo/servo/tree/master/tests/unit/script) and noticed it has it's own `Cargo.toml` file. So I wondering, could I override `rust_mozjs` in the `Cargo.toml` file to "patch" it (as in Python unit tests https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch) with my own package, which would live inside `/tests/unit/script`? That way I could just write a fake JS engine which would be called through `JS_WriteStructuredClone` and `JS_ReadStructuredClone` and which would invoke the callbacks, so I could actually unit test the entire flow of data. 

What do you think?

jdm commented:
I don't believe writing unit tests makes sense for this (and especially not using a different implementation of the underlying JS engine for that). Instead it makes sense to use [integration tests](https://github.com/servo/servo/blob/master/tests/wpt/README.md) that use web-facing APIs that make use of these callbacks. That requires integrating your work with an API like postMessage and writing a test that posts a blob between two iframes.

gterzian commented:
@jdm I'm wondering why the build is failing because the test manifest changed, I use `./mach create-wpt tests/wpt/mozilla/tests/mozilla/structuredclone.html` to do that, is there more to do in order to properly update the tests?

By the way, thanks a lot for the integration tests suggestion, I was able to first get my callback code to panic in a few different ways, and currently while it doesn't panic anymore, I am getting some weird values in the messages, so I think something is wrong with how I casted/transferred those values in between write and read callbacks(I am investigating and will shout if I get stuck). In other words, adding those tests was a great way to get some feedback on my running code 😄

gterzian commented:
@jdm nevermind I think I answered my own question with https://github.com/servo/servo/blob/master/docs/HACKING_QUICKSTART.md#add-a-new-test

ran a `./mach test-wpt tests/wpt/mozilla/tests/mozilla/structuredclone.html --manifest-update`

nox commented:
There are a bunch of `println!` statements that need to be removed.

jdm commented:
This is very much an ongoing WIP, so no need to specifically call out anything that isn't clearly wrong in the code.

gterzian commented:
Ok, this one is now ready for review 😄

gterzian commented:
@jdm Did you have a chance to take a look? Just pinging you again since I wasn't sure if you saw my request for review earlier 😄

jdm commented:
I'm on vacation right now, so @Ms2ger would be a more responsive reviewer.

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,124 @@
+#[allow(dead_code)]
+#[repr(u32)]
+enum StructuredCloneTags {
+    ScTagMin = 0xFFFF8000,
```
There is no need for the `ScTag` prefix.

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,124 @@
+            return *root.reflector().get_jsobject().ptr
+        }
+    }
+    return Heap::default().get()
```
That's an odd way to write `ptr::null_mut()`.

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,124 @@
+        let vec_mut_ptr: *mut raw::c_void = empty_vec.as_mut_ptr() as *mut _ as *mut raw::c_void;
+        mem::forget(empty_vec);
+        if JS_ReadBytes(r, vec_mut_ptr, data as usize) {
+            let blob_bytes: Vec<u8> = Vec::from_raw_parts(vec_mut_ptr as *mut u8, data as usize, data as usize);
```
You're leaking the vector if `JS_ReadBytes` fails.

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,124 @@
+                                   data: u32,
+                                   _closure: *mut raw::c_void) -> *mut JSObject {
+    if tag == StructuredCloneTags::ScTagDomBlob as u32 {
+        let mut empty_vec: Vec<u8> = Vec::with_capacity(data as usize);
```
```rust
let length = data as usize;
let mut buffer = vec![0u8; length];
if JS_ReadBytes(r, buffer.as_mut_ptr() as *mut c_void, length) {
```

should work out fine

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,124 @@
+        mem::forget(empty_vec);
+        if JS_ReadBytes(r, vec_mut_ptr, data as usize) {
+            let blob_bytes: Vec<u8> = Vec::from_raw_parts(vec_mut_ptr as *mut u8, data as usize, data as usize);
+            let js_context = GlobalScope::from_context(cx);
```
Please use `target_global` rather than `js_context`.

Ms2ger replied:
This is only correct if we have a `JSAutoCompartment` in place before calling `JS_ReadStructuredClone`.

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,124 @@
+            /// NOTE: missing the content-type string here.
+            /// Use JS_WriteString in write_callback? Make Blob.type_string pub?
+            let root = Blob::new(&js_context, BlobImpl::new_from_bytes(blob_bytes), "".to_string());
+            return *root.reflector().get_jsobject().ptr
```
```rust
return root.reflector().get_jsobject().get();
```

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,124 @@
+            let js_context = GlobalScope::from_context(cx);
+            /// NOTE: missing the content-type string here.
+            /// Use JS_WriteString in write_callback? Make Blob.type_string pub?
+            let root = Blob::new(&js_context, BlobImpl::new_from_bytes(blob_bytes), "".to_string());
```
`blob` rather than `root`, please.

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,124 @@
+                                    w: *mut JSStructuredCloneWriter,
+                                    obj: HandleObject,
+                                    _closure: *mut raw::c_void) -> bool {
+    if let Ok(blob) = root_from_handleobject::<Blob>(obj) {
```
I think this will be more maintainable going forward if we create a `write_blob` function and move the contents of this outer `if` block into it.

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,124 @@
+                                    _closure: *mut raw::c_void) -> bool {
+    if let Ok(blob) = root_from_handleobject::<Blob>(obj) {
+        if let Ok(blob_vec) = blob.get_bytes() {
+            let mut blob_vec_copy = blob_vec.to_vec();
```
No need for this copy; `get_bytes()` already returns a copy.

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,124 @@
+                                  StructuredCloneTags::ScTagDomBlob as u32,
+                                  blob_vec_copy_len as u32) {
+                let blob_vec_ptr: *const raw::c_void = blob_vec_copy.as_ptr() as *const _ as *const raw::c_void;
+                mem::forget(blob_vec_copy);
```
`JS_WriteBytes` does not take ownership of the buffer, so this is a memory leak. The good news is that you don't need to do this dance at all.

Ms2ger commented on tests/wpt/mozilla/tests/mozilla/structuredclone.html:
```diff
@@ -0,0 +1,32 @@
+<body>
+<script>
+async_test(function(t) {
+  var blob = new Blob(['<a id="a"><b id="b">hey!</b></a>'], {type:"text/plain"});
```
1. Why can't this test be submitted into web-platform-tests
2. Is there no test for this yet?

gterzian replied:
1. I don't know. I'm under the impression that what we're doing is Spidermonkey specific, so I thought those tests belonged in the Mozilla folder.
2. I couldn't find a WPT test specifically testing structured cloning of blobs. I found tests for  [structured cloning image data](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/workers/interfaces/DedicatedWorkerGlobalScope/postMessage/structured-clone-imagedata.html) and [basic JS types](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/workers/interfaces/DedicatedWorkerGlobalScope/postMessage/structured-clone-message.html). 
There is also one test that uses [`postMessage` with a blob](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/offscreen-canvas/the-offscreen-canvas/offscreencanvas.convert.to.blob.w.html#L34) in `offscreen-canvas`.

Ms2ger replied:
Let's move it to `tests/wpt/web-platform-tests/html/infrastructure/safe-passing-of-structured-data`.

Ms2ger reviewed (changes requested):
Initial comments from a first read-through. General design looks sound, so please don't be discouraged by the comments below :)

gterzian commented:
@Ms2ger Thanks for the review, it is quite the opposite to discouraging, I was doing quite a bit of guesswork there so I'm happy with the guidance 👍 

Could you take another look please?

gterzian commented:
Any ideas why the build is failing with: `Running `rustc --crate-name servo main.rs (...)`
`error: linking with `cc` failed: exit code: 1`

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,129 @@
+///TODO: Add Min and Max const to https://github.com/servo/rust-mozjs/blob/master/src/consts.rs?
+///TODO: Determine which value Min and Max should have.
+///NOTE: Current values found at https://dxr.mozilla.org/mozilla-central/source/js/public/StructuredClone.h#323
+#[allow(dead_code)]
```
I don't think anything here actually is dead code. Can you remove those annotations?

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,129 @@
 
+///TODO: Add Min and Max const to https://github.com/servo/rust-mozjs/blob/master/src/consts.rs?
+///TODO: Determine which value Min and Max should have.
+///NOTE: Current values found at https://dxr.mozilla.org/mozilla-central/source/js/public/StructuredClone.h#323
```
Those notes/todos should be normal comments (`//`), not documentation comments (`///`).

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,129 @@
+
+unsafe extern "C" fn read_blob(cx: *mut JSContext,
+                               r: *mut JSStructuredCloneReader,
+                               data: u32) -> *mut JSObject {
```
For multiline argument lists, we prefer putting the return type on its own line too:
```rust

unsafe extern "C" fn read_blob(cx: *mut JSContext,
                               r: *mut JSStructuredCloneReader,
                               data: u32)
                               -> *mut JSObject {
```

Same comment a few more times below.

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,129 @@
+unsafe extern "C" fn write_blob(blob: Root<Blob>,
+                                w: *mut JSStructuredCloneWriter) -> bool {
+    if let Ok(mut blob_vec) = blob.get_bytes() {
+        blob_vec.shrink_to_fit();
```
I'm not sure I see the point of this call.

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,129 @@
+        let length = blob_vec.len();
+        assert!(JS_WriteUint32Pair(w,
+                                   StructuredCloneTags::DomBlob as u32,
+                                   length as u32));
```
I just realized now that this will truncate a blob with length greater than 2**32. We should probably pass 0 here, and call `JS_WriteUint32Pair` again with the length split up.

Ms2ger reviewed (changes requested):
Thanks! Some more comments below.

gterzian commented:
Thanks, all done. I also added an assertion in the read callback that the tag is within the expected range...

gterzian commented:
@Ms2ger did you have a chance to take another look?

I'm also wondering about the two TODO's I left in the code:

1. Do we need to move the `StructuredCloneTags:Min` and `StructuredCloneTags:Max` to the `rust-mozjs` api? @jdm suggested previously that they might need to be added as constants to https://github.com/servo/rust-mozjs/blob/master/src/consts.rs . Currently value I just copy pasted from https://dxr.mozilla.org/mozilla-central/source/js/public/StructuredClone.h#323
2. Do we need to clone the `content-type` string of the Blob? Currently the clone will have an empty `content-type` in all case. This is because the `content-type` on `Blob` isn't public at this point, and I left it as it was. Should we make it public, and clone it as well? In comparison, I don't see Gecko cloning the `content-string` either, see https://dxr.mozilla.org/mozilla-central/source/dom/base/StructuredCloneHolder.cpp#592 (They only store the `BlobImpl` in an array and read/write the index of it as `Uint32Pair`)

jdm commented:
Ah, Gecko's system works by storing pointers to objects that can be shared between threads and marking the blob contents as immutable. That's why they don't do any serialization of the blob contents or type. I wonder if that gets them the correct behaviour for File objects for free?

jdm commented:
Then again, there's also the code in the IndexedDB callbacks which writes out information like the content type explicitly: https://dxr.mozilla.org/mozilla-central/rev/ff04d410e74b69acfab17ef7e73e7397602d5a68/dom/indexedDB/IDBObjectStore.cpp#284-345

gterzian commented:
@jdm yes I initially also tried to just store the `Blob` somewhere, however I ran into lots of problems such as:
1. It seems Rust really doesn't like a module level static array(which would be available from both the read and write callback). 
2. I could instead turn the `StructuredCloneData` into a struct, with an array to store blobs, and pass it around as the `closure` argument when writing and reading. However in that case I think the problems related to the fact that Blob doesn't implement traits that would allow passing it around like that, I think it was mainly the `Send` trait. 

Personally I also feel like using the various `JS_*` functions and let the JS engine deal with the data is perhaps easier than introducing more mutable data on our side.

jdm commented:
Yeah, using a setup like Gecko's would require splitting the relevant fields out of Blob and into a separate BlobInternals type that could be stored in an Arc, and creating an enum that could either store BlobInternals or `Arc<BlobInternals>` depending on whether the blob is immutable or not. Just passing around `JS<Blob>` wouldn't work because those are JS objects that are associated with a particular JS runtime.

gterzian commented:
@jdm thanks for the info. I think we can first try to write/read the content string, and see where it takes us, while keeping a open mind for an approach closer to the one in Gecko if we identity a problem with the current one...

Ms2ger commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,131 @@
+                                   data: u32,
+                                   _closure: *mut raw::c_void)
+                                   -> *mut JSObject {
+    assert!(tag < StructuredCloneTags::Max as u32 && tag > StructuredCloneTags::Min as u32);
```
Nit: split into two asserts (I think there's `assert_gt` and `assert_lt` that would also make the error message better).

Ms2ger reviewed (commented):
No more comments from me; will let jdm finish.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,131 @@
+    assert!(JS_ReadBytes(r, buffer.as_mut_ptr() as *mut raw::c_void, length));
+    let target_global = GlobalScope::from_context(cx);
+    // NOTE: missing the content-type string here.
+    // Use JS_WriteString in write_callback? Make Blob.type_string pub?
```
We should write and read the type string. We'll need a public method on Blob that exposes it, rather than making the member public.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,131 @@
+
+unsafe extern "C" fn read_blob(cx: *mut JSContext,
+                               r: *mut JSStructuredCloneReader,
+                               _data: u32)
```
No need for this argument, since it's unused.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,131 @@
+    Max = 0xFFFFFFFF,
+}
+
+unsafe extern "C" fn read_blob(cx: *mut JSContext,
```
No need for `extern "C"` here, since this isn't called directly by C code.

jdm commented on tests/wpt/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/structured_clone_blob.html:
```diff
@@ -0,0 +1,32 @@
+<!doctype html>
+<meta charset="utf-8">
+<title></title>
```
We should add a title to this test.

jdm replied:
`<meta>` and `<title>` belong in the `<head>` block.

jdm commented on tests/wpt/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/structured_clone_blob.html:
```diff
@@ -0,0 +1,32 @@
+          var reader2 = new FileReader();
+          reader2.addEventListener("loadend", this.step_func_done(function() {
+              cloned_content = reader2.result;
+              assert_equals(cloned_content, original_content);
```
Let's assert the types match, too.

jdm commented on tests/wpt/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/structured_clone_blob.html:
```diff
@@ -0,0 +1,32 @@
+      reader.readAsText(blob);
+  }), false);
+  window.postMessage(blob, '*');
+}, "test cloning a Blob");
```
`"Cloning a Blob into the same realm"`

jdm commented on tests/wpt/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/structured_clone_blob.html:
```diff
@@ -0,0 +1,32 @@
+async_test(function(t) {
+  var blob = new Blob(['<a id="a"><b id="b">hey!</b></a>'], {type:"text/plain"});
+  window.addEventListener("message", this.step_func(function(msg) {
+      assert_equals(msg.data.size, blob.size);
```
Add a `assert_true(msg.data instanceof Blob)`, too.

gterzian replied:
this was done as well, although two lines lower so Github doesn't show it as outdated...

jdm replied:
It probably makes sense to check that it's a Blob first, otherwise this check is likely to fail in a slightly more confusing fashion.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,131 @@
 
+// TODO: Should we add Min and Max const to https://github.com/servo/rust-mozjs/blob/master/src/consts.rs?
+// TODO: Determine for sure which value Min and Max should have.
+// NOTE: Current values found at https://dxr.mozilla.org/mozilla-central/source/js/public/StructuredClone.h#323
```
https://dxr.mozilla.org/mozilla-central/rev/ff04d410e74b69acfab17ef7e73e7397602d5a68/js/public/StructuredClone.h#323 is a more permanent link.

jdm reviewed (changes requested):
This is great work! Thank you for tackling and exploring it!

gterzian commented:
@jdm thanks! Latest comments addressed, I've added a `get_type_string` method on `Blob` which returns a `Vec<u8>`, and now also writing/reading those as bytes (I found out the `JS_WriteString` function probably has a different intent from what I expected, as there also is no corresponding `JS_ReadString` function to be found, neither in `rust-mozjs` or Gecko.)

I've also added an assertion in the wpt test regarding the type string.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,148 @@
+    assert!(JS_ReadBytes(r, blob_buffer.as_mut_ptr() as *mut raw::c_void, blob_length));
+    let mut type_str_buffer = vec![0u8; type_str_length];
+    assert!(JS_ReadBytes(r, type_str_buffer.as_mut_ptr() as *mut raw::c_void, type_str_length));
+    let type_str = String::from_raw_parts(type_str_buffer.as_ptr() as *mut _, type_str_length, type_str_length);
```
String::from_raw_parts takes ownership of the pointer; we should be using String::from_utf8_unchecked on the vector instead.

jdm commented on components/script/dom/bindings/structuredclone.rs:
```diff
@@ -5,15 +5,148 @@
+
+unsafe fn write_blob(blob: Root<Blob>,
+                     w: *mut JSStructuredCloneWriter)
+                     -> bool {
```
Let's make this return `Result<(), ()>` for clarity. Then we can do:
```rust
let blob_vec = try!(blob.get_bytes());
...
return Ok(());
```

gterzian replied:
thanks, I was about to ask you if the code needed to do anything in case of an error, and then read about the early return of `try!`, interesting macro that is...

jdm commented on components/script/dom/blob.rs:
```diff
@@ -166,6 +166,11 @@ impl Blob {
     }
 
+    /// Get a copy of the type_string
+    pub fn get_type_string(&self) -> Vec<u8> {
```
Let's make this return the string value, and have the caller convert it into a `Vec<u8>` instead.

jdm commented:
I've opened https://github.com/servo/servo/issues/16031 for some improvements we can make once this merges :)

gterzian commented:
@jdm thanks again for the pointers, all comments have been addressed...

jdm commented:
@bors-servo: r+

bors-servo commented:
:clipboard: Looks like this PR is still in progress, ignoring approval

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 2b45871 has been approved by `jdm`

<!-- @bors-servo r=jdm 2b45871f9695b199fa0051c89c38f6159d83cd73 -->

bors-servo commented:
:hourglass: Testing commit 2b45871f9695b199fa0051c89c38f6159d83cd73 with merge 17aa7560818914b46ba78f1455115aa645462180...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/5869)

jdm commented:
```
warning: literal out of range for isize
  --> /home/servo/buildbot/slave/android/build/components/script/dom/bindings/structuredclone.rs:32:11
   |
32 |     Min = 0xFFFF8000,
   |           ^^^^^^^^^^
   |
   = note: #[warn(overflowing_literals)] on by default

warning: literal out of range for isize
  --> /home/servo/buildbot/slave/android/build/components/script/dom/bindings/structuredclone.rs:33:15
   |
33 |     DomBlob = 0xFFFF8001,
   |               ^^^^^^^^^^
   |
   = note: #[warn(overflowing_literals)] on by default

warning: literal out of range for isize
  --> /home/servo/buildbot/slave/android/build/components/script/dom/bindings/structuredclone.rs:34:11
   |
34 |     Max = 0xFFFFFFFF,
   |           ^^^^^^^^^^
   |
   = note: #[warn(overflowing_literals)] on by default

error: bitshift exceeds the type's number of bits
  --> /home/servo/buildbot/slave/android/build/components/script/dom/bindings/structuredclone.rs:39:19
   |
39 |   let high: u32 = (length >> 32) as u32;
   |                   ^^^^^^^^^^^^^^
   |
   = note: #[deny(exceeding_bitshifts)] on by default

warning: unused import: `servo_config::opts`
  --> /home/servo/buildbot/slave/android/build/components/script/dom/permissions.rs:20:5
   |
20 | use servo_config::opts;
   |     ^^^^^^^^^^^^^^^^^^
   |
   = note: #[warn(unused_imports)] on by default

error: aborting due to previous error
```
We'll need to write the code dealing with usize values in ways that support both 32 bit and 64 bit values.

gterzian commented:
@jdm ok! Am I correct to think that the high/low dance regarding usize is only necessary on 64 bits architecture? Regarding the tags, frankly I just copy/pasted the values from Gecko, so I am unsure how to adapt them. Is it very naive to think we should halve them on 34 bits?

jdm commented:
Ah, the issue is that the default enum representation is isize, and 0xFFFF8000 is out of range of i32. It fits in u32, however, and that's the type of the original tag: https://dxr.mozilla.org/mozilla-central/rev/05bfa2831c0ba4a26fa72328ffe6a99aba9c356a/js/public/StructuredClone.h#324

We should do this instead of providing separate values for 32/64:
```rust
#[repr(u32)]
enum StructuredCloneTags {
    Min = 0xFFFF8000,
    DomBlob = 0x7FFFC001,
    Max = 0xFFFFFFFF,
}
```

jdm commented:
The other changes in that commit look fine, however.

gterzian commented:
@jdm thanks! I have made the suggested change, and squashed as well...

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 5c4f0be has been approved by `jdm`

<!-- @bors-servo r=jdm 5c4f0be04872f5f2e8736b15a2169f9479509a0e -->

bors-servo commented:
:hourglass: Testing commit 5c4f0be04872f5f2e8736b15a2169f9479509a0e with merge 07e3a63b4259612abba4b7526d4807850483bbdc...

Manishearth commented:
@bors-servo r- r=jdm retry clean

(need to prioritize stylo PRs that need autoland fiddling to land)

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

bors-servo commented:
:pushpin: Commit 5c4f0be has been approved by `jdm`

<!-- @bors-servo r=jdm 5c4f0be04872f5f2e8736b15a2169f9479509a0e -->

bors-servo commented:
:hourglass: Testing commit 5c4f0be04872f5f2e8736b15a2169f9479509a0e with merge 7be9f0e7c4f16c5626180c3a6b07abdecea408fe...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/5899), [arm32](http://build.servo.org/builders/arm32/builds/6122), [arm64](http://build.servo.org/builders/arm64/builds/6154), [linux-dev](http://build.servo.org/builders/linux-dev/builds/6260), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/2974), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3001), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6263), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/6290), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/2938), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3013), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/1427)
Approved by: jdm
Pushing 7be9f0e7c4f16c5626180c3a6b07abdecea408fe to master...

gterzian commented:
nice one! Thank you for your help @Ms2ger and @jdm
